crates/neon/src/types_impl/buffer/types.rs

```diff
@@ -0,0 +1,721 @@
+    pub fn from_buffer<'cx, C>(cx: &mut C, buffer: Handle<JsArrayBuffer>) -> JsResult<'cx, Self>
+    where
+        C: Context<'cx>,
+    {
+        let byte_length = buffer.byte_length(cx);
+        let elt_size = std::mem::size_of::<T>();
+        let len = byte_length / elt_size;
+
+        if (len * elt_size) != byte_length {
+            panic!(
+                "byte length of typed array should be a multiple of {}",
+                elt_size
+            );
+        }
+
+        Self::from_buffer_region(cx, buffer, 0, len)
+    }
+
+    /// Constructs a typed array that views a region of `buffer` starting
+    /// at the specified byte offset and with the specified number of elements.
+    ///
+    /// The resulting typed array has `len` elements and byte length
+    /// `(len * size_of::<T>())`.
+    pub fn from_buffer_region<'cx, C>(
+        cx: &mut C,
+        buffer: Handle<JsArrayBuffer>,
+        byte_offset: usize,
+        len: usize,
+    ) -> JsResult<'cx, Self>
+    where
+        C: Context<'cx>,
+    {
+        let result = unsafe {
+            sys::typedarray::new(
+                cx.env().to_raw(),
+                T::TYPE_TAG,
+                buffer.to_raw(),
+                byte_offset,
+                len,
+            )
+        };
+
+        if let Ok(arr) = result {
+            Ok(Handle::new_internal(Self {
+                local: arr,
+                _type: PhantomData,
+            }))
+        } else {
+            Err(Throw::new())
+        }
+    }
+
+    /// Constructs a new typed array of length `len`.
+    ///
+    /// The resulting typed array has a newly allocated storage buffer of
+    /// size `(len * size_of::<T>())` bytes.
+    pub fn new<'cx, C>(cx: &mut C, len: usize) -> JsResult<'cx, Self>
+    where
+        C: Context<'cx>,
+    {
+        let buffer = cx.array_buffer(len * std::mem::size_of::<T>())?;
+        Self::from_buffer_region(cx, buffer, 0, len)
+    }
+
+    /// Returns the [`JsArrayBuffer`](JsArrayBuffer) that owns the underlying storage buffer
+    /// for this typed array.
+    ///
+    /// Note that the typed array may only reference a region of the buffer; use the
+    /// [`byte_offset()`](JsTypedArray::byte_offset) and
+    /// [`byte_length()`](crate::types::buffer::TypedArray::byte_length) methods to
+    /// determine the region.
+    pub fn buffer<'cx, C>(&self, cx: &mut C) -> Handle<'cx, JsArrayBuffer>
+    where
+        C: Context<'cx>,
+    {
+        let env = cx.env();
+        let info = unsafe { sys::typedarray::info(env.to_raw(), self.to_raw()) };
+        Handle::new_internal(JsArrayBuffer::from_raw(env, info.buf))
+    }
+
+    /// Returns the offset (in bytes) of the typed array from the start of its
+    /// [`JsArrayBuffer`](JsArrayBuffer).
+    pub fn byte_offset<'cx, C>(&self, cx: &mut C) -> usize
+    where
+        C: Context<'cx>,
+    {
+        let env = cx.env();
+        let info = unsafe { sys::typedarray::info(env.to_raw(), self.to_raw()) };
```

Comment:
It's not very efficient to repeatedly call `typedarray::info` for this data. Can we memoize it on an `Option` field in `JsTypedArray`?

It does make the struct a little larger to copy, but I think that's okay since it's still not *that* large.

Reply:
I like this idea but didn't we run into trouble before with making some handles into fat pointers?

Reply:
No, we have a couple. Most notably is `JsBox`.

Reply:
OK, I pushed an implementation of this optimization! See what you think.